src/imitation/rewards/reward_nets.py

```diff
@@ -550,3 +574,114 @@ def __init__(
+    @abc.abstractmethod
+    def forward(
+        self,
+        state: th.Tensor,
+        action: th.Tensor,
+        next_state: th.Tensor,
+        done: th.Tensor,
+    ) -> th.Tensor:
+        """Compute rewards for a batch of transitions and keep gradients."""
+        rewards = []
+        for net in self.base_networks:
+            rewards.append(net(state, action, next_state, done))
+        rewards = th.stack(rewards, dim=0)
+        return rewards.mean(0)
+
+    @th.no_grad()
+    def standard_deviation(
+        self,
+        state: np.ndarray,
+        action: np.ndarray,
+        next_state: np.ndarray,
+        done: np.ndarray,
+    ) -> np.ndarray:
+        """Compute the standard deviation of the reward distribution for a batch.
+
+        Args:
+            state: Current states of shape `(batch_size,) + state_shape`.
+            action: Actions of shape `(batch_size,) + action_shape`.
+            next_state: Successor states of shape `(batch_size,) + state_shape`.
+            done: End-of-episode (terminal state) indicator of shape `(batch_size,)`.
+
+        Returns:
+            Computed reward std of shape `(batch_size,)`.
+        """
+        state, action, next_state, done = self.preprocess(
+            state,
+            action,
+            next_state,
+            done,
+        )
+        rewards = []
+        for net in self.base_networks:
+            rewards.append(net(state, action, next_state, done))
+        rewards = th.stack(rewards, dim=0)
+        return rewards.std(0)
+
+
+class ConservativeRewardWrapper(RewardNetWrapper):
+    """A reward network that returns a conservative estimate of the reward."""
+
+    base: UncertainRewardNet
+
+    def __init__(self, base: UncertainRewardNet, alpha: float = 1.0):
+        """Create a conservative reward network.
+
+        Args:
+            base: An uncertain rewarard network
+            alpha: multiple of standard deviation to subtract from reward mean when
+                predicting conservative reward. Defaults to 1.0.
+        """
+        super().__init__(base)
+        self.alpha = alpha
+
+    def predict_processed(
+        self,
+        state: np.ndarray,
+        action: np.ndarray,
+        next_state: np.ndarray,
+        done: np.ndarray,
+    ) -> np.ndarray:
+        """Compute a lower confidence bound on the reward without gradients.
+
+        Args:
+            state: Current states of shape `(batch_size,) + state_shape`.
+            action: Actions of shape `(batch_size,) + action_shape`.
+            next_state: Successor states of shape `(batch_size,) + state_shape`.
+            done: End-of-episode (terminal state) indicator of shape `(batch_size,)`.
+
+        Returns:
+            Computed lower confidence bounds on rewards of shape `(batch_size,`).
+        """
+        reward_mean = self.base.predict_processed(state, action, next_state, done)
+        reward_std = self.base.predict_processed(state, action, next_state, done)
```

Comment:
```suggestion
        reward_std = self.base.standard_deviation(state, action, next_state, done)
```
?

Reply:
Also a bit inefficient (doing forward pass twice, once for `reward_mean` and once for `reward_std`)